lib/AST/DiagnosticConsumer.cpp

```diff
@@ -163,41 +157,75 @@ void FileSpecificDiagnosticConsumer::handleDiagnostic(
     const DiagnosticInfo &Info) {
 
-  Optional<DiagnosticConsumer *> specificConsumer;
+  HasAnErrorBeenConsumed |= Kind == DiagnosticKind::Error;
+
+  Optional<ConsumerSpecificInformation *> consumerSpecificInfo;
   switch (Kind) {
   case DiagnosticKind::Error:
   case DiagnosticKind::Warning:
   case DiagnosticKind::Remark:
-    specificConsumer = consumerForLocation(SM, Loc);
-    ConsumerForSubsequentNotes = specificConsumer;
+    consumerSpecificInfo = consumerSpecificInformationForLocation(SM, Loc);
+    ConsumerSpecificInfoForSubsequentNotes = consumerSpecificInfo;
     break;
   case DiagnosticKind::Note:
-    specificConsumer = ConsumerForSubsequentNotes;
+    consumerSpecificInfo = ConsumerSpecificInfoForSubsequentNotes;
     break;
   }
-
-  if (!specificConsumer.hasValue()) {
+  if (!consumerSpecificInfo.hasValue()) {
     for (auto &subConsumer : SubConsumers) {
       if (subConsumer.second) {
         subConsumer.second->handleDiagnostic(SM, Loc, Kind, FormatString,
                                              FormatArgs, Info);
       }
     }
-  } else if (DiagnosticConsumer *c = specificConsumer.getValue())
-    c->handleDiagnostic(SM, Loc, Kind, FormatString, FormatArgs, Info);
-  else
-    ; // Suppress non-primary diagnostic in batch mode.
+    return;
+  }
+  if (!consumerSpecificInfo.getValue()->consumer)
+    return; // Suppress non-primary diagnostic in batch mode.
+
+  consumerSpecificInfo.getValue()->consumer->handleDiagnostic(
+      SM, Loc, Kind, FormatString, FormatArgs, Info);
+  consumerSpecificInfo.getValue()->hasAnErrorBeenEmitted |=
+      Kind == DiagnosticKind::Error;
 }
 
-bool FileSpecificDiagnosticConsumer::finishProcessing() {
+bool FileSpecificDiagnosticConsumer::finishProcessing(SourceManager &SM) {
+  addNonSpecificErrors(SM);
+
   // Deliberately don't use std::any_of here because we don't want early-exit
   // behavior.
+
   bool hadError = false;
   for (auto &subConsumer : SubConsumers)
-    hadError |= subConsumer.second && subConsumer.second->finishProcessing();
+    hadError |= subConsumer.second && subConsumer.second->finishProcessing(SM);
   return hadError;
 }
 
+static void produceNonSpecificError(DiagnosticConsumer *consumer,
+                                    SourceManager &SM) {
+  Diagnostic diagnostic(diag::error_an_error_occurred);
+
+  // Stolen from DiagnosticEngine::emitDiagnostic
+  DiagnosticInfo Info;
+  Info.ID = diagnostic.getID();
+
+  consumer->handleDiagnostic(
+      SM, SourceLoc(), DiagnosticKind::Error,
+      DiagnosticEngine::diagnosticStringFor(diagnostic.getID()), {}, Info);
+}
+
+void FileSpecificDiagnosticConsumer::addNonSpecificErrors(SourceManager &SM) {
+  if (!HasAnErrorBeenConsumed)
+    return;
+  for (auto &info : ConsumersOrderedByRange) {
+    if (!info.hasAnErrorBeenEmitted && info.consumer) {
+      produceNonSpecificError(info.consumer, SM);
+      info.hasAnErrorBeenEmitted = true;
+      HasAnErrorBeenConsumed = true;
```

Comment:
This line should be unnecessary since `HasAnErrorBeenConsumed` is guarding this section.

Reply:
Thanks. I must be losing focus.